examples/server/server.cpp

```diff
@@ -554,13 +328,10 @@ struct llama_server_context
     // slots / clients
     std::vector<llama_client_slot> slots;
 
-    std::vector<task_server> queue_tasks;
-    std::vector<task_result> queue_results;
+    llama_server_queue<task_server> queue_tasks;
+    llama_server_response_event queue_results;
     std::vector<task_multi>  queue_multitasks;
-    std::mutex mutex_tasks; // also guards id_gen, and queue_multitasks
-    std::condition_variable condition_tasks;
-    std::mutex mutex_results;
-    std::condition_variable condition_results;
+    std::mutex mutex_multitasks;
```

Comment:
Refactoring all synchronization and queues into `utils.hpp` is OK, but the result should be that there are no longer any mutexes or locks in `server.cpp`

Can we avoid this somehow?

Reply:
Sure, and I suspect that `queue_multitasks` may not even need mutex.

The reason why we needed mutex was because `queue_multitasks` is modified by both main thread and http thread (by calling `llama.next_result`)

But my new implementation of `llama.next_result` (moved to `llama_server_response_event.recv`) does not touch `queue_multitasks` anymore. Instead, the multitask queue is updated via `llama_server_response_event.send`, which is always called by main thread.

But ideally, I want multitask & task to be in the same place, i.e. maybe add it to the new `struct llama_server_queue`

Reply:
multitasks is moved into `llama_server_queue` in https://github.com/ggerganov/llama.cpp/pull/5065/commits/d87b48fd55276f4996caa9149855e82f7acdbff8 and https://github.com/ggerganov/llama.cpp/pull/5065/commits/1bd867894d648325e5e976382c95f2af2bad7688

I didn't have many idea to test it though, but seems to work just fine:

---

How I tested it: Send POST to `http://localhost:8080/completion`

Body:

```
{
    "prompt": ["hi", "bonjour"],
    "n_predict": 10
}
```

Response (expected 1 response in english and 1 in french):

```
{
    "results": [
        {
            "content": ", 2018)\n\nSangam (2019",
            ...
        },
        {
            "content": " le chôme-Ouest, b",
            ...
        }
    ]
}
```

---

Another test using OAI `/v1/chat/completions` and `stream: true` also works fine.

I didn't have time to test embedding API, but I believe that it should work.